libraries/chain/controller.cpp

```diff
@@ -4143,65 +4157,34 @@ struct controller_impl {
          assert(!verify_qc_future.valid());
       }
 
+      bool best_head = forkdb.add(bsp, ignore_duplicate_t::yes);
+      if constexpr (is_proper_savanna_block)
+         vote_processor.notify_new_block(async_aggregation);
 
-      if (!should_terminate(bsp->block_num())) {
-         forkdb.add(bsp, ignore_duplicate_t::yes);
-         if constexpr (is_proper_savanna_block)
-            vote_processor.notify_new_block(async_aggregation);
-      }
-
-      return block_handle{bsp};
-   }
-
-   std::future<block_handle> create_block_handle_future( const block_id_type& id, const signed_block_ptr& b ) {
-      EOS_ASSERT( b, block_validate_exception, "null block" );
-
-      auto f = [&](auto& forkdb) -> std::future<block_handle> {
-         return post_async_task( thread_pool.get_executor(), [b, id, &forkdb, control=this]() {
-            auto prev = forkdb.get_block( b->previous, include_root_t::yes );
-            EOS_ASSERT( prev, unlinkable_block_exception,
-                        "unlinkable block ${id} previous ${p}", ("id", id)("p", b->previous) );
-
-            return control->create_block_state_i( forkdb, id, b, *prev );
-         } );
-      };
-
-      // always return a valid future
-      auto unlinkable = [&](const auto&) -> std::future<block_handle> {
-         std::packaged_task<block_handle()> task( [b, id]() -> block_handle {
-            EOS_ASSERT( false, unlinkable_block_exception,
-                        "unlinkable block ${id} previous ${p} not in fork db", ("id", id)("p", b->previous) );
-         } );
-         task();
-         return task.get_future();
-      };
-
-      if (!b->is_proper_svnn_block()) {
-         return fork_db.apply<std::future<block_handle>>(f, unlinkable);
-      }
-      return fork_db.apply<std::future<block_handle>>(unlinkable, f);
+      return std::tuple{best_head, block_handle{std::move(bsp)}};
    }
 
    // thread safe, expected to be called from thread other than the main thread
-   std::optional<block_handle> create_block_handle( const block_id_type& id, const signed_block_ptr& b ) {
+   controller::accepted_block_handle create_block_handle( const block_id_type& id, const signed_block_ptr& b ) {
       EOS_ASSERT( b, block_validate_exception, "null block" );
       
-      auto f = [&](auto& forkdb) -> std::optional<block_handle> {
-         // previous not found could mean that previous block not applied yet
+      auto f = [&](auto& forkdb) -> controller::accepted_block_handle {
+         // previous not found, means it is unlinkable
          auto prev = forkdb.get_block( b->previous, include_root_t::yes );
          if( !prev ) return {};
 
-         return create_block_state_i( forkdb, id, b, *prev );
+         auto [best_head, obh] = create_block_state_i( forkdb, id, b, *prev );
```

Comment:
Why not have `create_block_state_i` return a `controller::accepted_block_handle`?

Reply:
Not sure how I missed that.